xls/contrib/xlscc/expr_clone.cc

```
@@ -0,0 +1,261 @@
+        expr->getValueKind(), expr->getObjectKind(), expr->getRBracketLoc());
+  }
+
+  clang::Expr* VisitInitListExpr(const clang::InitListExpr* expr) {
+    std::vector<clang::Expr*> inits;
+    for (auto* init : expr->inits()) {
+      inits.push_back(Visit(init));
+    }
+    auto* cloned = new (ctx) clang::InitListExpr(ctx, expr->getLBraceLoc(),
+                                                 inits, expr->getRBraceLoc());
+    cloned->setType(expr->getType());
+    return cloned;
+  }
+
+  clang::Expr* VisitCXXOperatorCallExpr(
+      const clang::CXXOperatorCallExpr* expr) {
+    std::vector<clang::Expr*> args;
+    for (auto* arg : expr->arguments()) {
+      args.push_back(Visit(arg));
+    }
+    return clang::CXXOperatorCallExpr::Create(
+        ctx, expr->getOperator(), Visit(expr->getCallee()), args,
+        expr->getType(), expr->getValueKind(), expr->getRParenLoc(),
+        expr->getFPFeatures());
+  }
+
+  clang::Expr* VisitCXXMemberCallExpr(const clang::CXXMemberCallExpr* expr) {
+    std::vector<clang::Expr*> args;
+    for (auto* arg : expr->arguments()) {
+      args.push_back(Visit(arg));
+    }
+    return clang::CXXMemberCallExpr::Create(
+        ctx, Visit(expr->getCallee()), args, expr->getType(),
+        expr->getValueKind(), expr->getRParenLoc(), expr->getFPFeatures());
+  }
+
+  clang::Expr* VisitCXXTemporaryObjectExpr(
+      const clang::CXXTemporaryObjectExpr* expr) {
+    std::vector<clang::Expr*> args;
+    for (auto* arg : expr->arguments()) {
+      args.push_back(Visit(arg));
+    }
+    return clang::CXXTemporaryObjectExpr::Create(
+        ctx, expr->getConstructor(), expr->getType(), expr->getTypeSourceInfo(),
+        args, expr->getParenOrBraceRange(), expr->hadMultipleCandidates(),
+        expr->isListInitialization(), expr->isStdInitListInitialization(),
+        expr->requiresZeroInitialization());
+  }
+
+  clang::Expr* VisitExpr(const clang::Expr* expr) { return nullptr; }
```

Could you store the specific clang::Expr* here for error reporting?

---

I don't think `StatusOr` can hold arbitrary data, but now I'm storing the name of the offending expression here which should be good enough for error reporting.